Add P-tuning for few_shot

tianxin1860 opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/PaddleNLP/pull/26 -->
### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
New features 
### PR changes
<!-- One of [ Models | APIs | Docs | Others ] -->
Models
### Description
<!-- Describe what this PR does -->
1. Add P-tuning strategy
2. Evaluate p-tuning strategy on FewClue dataset (9/9)
3. Add FewCLUE datasets to PaddleNLP

Todo:
1. Add PET strategy
2. Add EFL strategy
3. Add predict for FewCLUE private_test.json

Commit:
"add mlm params to dygraph ernie1.0"

Commit:
finish p-tuning v1.0

Commit:
mend

Commit:
delete unused coment

Commit:
Merge branch 'develop' into few_shot

Commit:
add label_normalized

Commit:
P-tuning: support Chid task of FewCLUE

ZeyuChen commented on examples/few_shot/p-tuning/ptuning_chid.py:
```diff
@@ -0,0 +1,293 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
CHID特殊处理是因为什么原因？

tianxin1860 replied:
常规任务转换成 MLM 任务或者生成任务之后，feature 内只需要抽象出 4 个 slot `[src_ids,  token_type_ids, masked_positions, masked_lm_labels]` 即可，但是 CHID 任务是在给定的 7 个候选成语中选择正确答案，生成的 feature 会有 5 个 slot, 多出 1 个 `candidate_label_ids`, 这个特殊性导致 CHID 任务的评估逻辑也没法和常规任务共用 1 个评估函数。和 @guoshengCS 讨论了下，会把评估逻辑从入口脚本抽离出来，每个任务根据任务名映射到各自的评估函数，主训练逻辑所有任务共用 1 套脚本。

tianxin1860 replied:
`CHID` 任务和其它任务训练脚本已经统一为 `p-tuning.py`,  thanks.

ZeyuChen commented on examples/few_shot/p-tuning/ptuning_chid.py:
```diff
@@ -0,0 +1,293 @@
+
+            # [bathc_size, label_length(4)] 
+            single_candidate_label_ids = candidate_label_ids[:, label_idx, :]
+            # 计算每个候选 label 的联合概率分布
```
`Calculate joint probability distribution for each candidate labels`
使用英文注释

tianxin1860 replied:
好的，thanks

tianxin1860 replied:
已修正，thanks.

ZeyuChen commented on examples/few_shot/p-tuning/data.py:
```diff
@@ -0,0 +1,438 @@
+            origin_label = example['label_desc']
+
+            if origin_label in label_normalize_dict:
+                # 针对部分 Label 进行标准化， eg. 英文 -> 中文
```
Normalize some of the labels, eg. ...

tianxin1860 replied:
好的，thanks

tianxin1860 replied:
已修正，thanks.

ZeyuChen commented on examples/few_shot/p-tuning/data.py:
```diff
@@ -0,0 +1,438 @@
+                # filter illegal example
+                continue
+
+            example["sentence1"] = example["sentence"]
```
看是否可以fewshot dataset 下沉，然后通过transform等on the fly操作通过map函数来处理？
这里确实有一点灵活性和可读性之间的平衡

tianxin1860 replied:
好的

tianxin1860 replied:
FewCLUE 的 9 个数据集已经集成到了 `fewclue.py` 中，目前的策略实现是针对每个 `xxx task` 有 1个用户自定义的 `transform_xxx` 函数来控制任务转换逻辑.

ZeyuChen commented on examples/few_shot/p-tuning/data.py:
```diff
@@ -0,0 +1,438 @@
+            origin_label = example["label"]
+
+            if origin_label in label_normalize_dict:
+                # 针对部分 Label 进行标准化， eg. 英文 -> 中文
```
我理解这部分normalize的操作就属于策略的一部分

tianxin1860 replied:
是的，label_normalize 属于策略的一部分，这块逻辑的确需要和 IO  解耦.

tianxin1860 replied:
已解耦，thanks.

ZeyuChen commented on examples/few_shot/p-tuning/data.py:
```diff
@@ -0,0 +1,438 @@
+    return examples
+
+
+def read_cluewsc(data_path, label_normalize_dict=None):
```
这种写法有他的好处，确实可读性会更加简单清晰。所以这里我觉得不太好强制，可能需要你再平衡下。
我理解数据集都不会太大，可以通过load_dataset的方法完整加载玩数据集再内存中。
然后通过这个策略方法遍历一遍数据集完成特征处理和normalize。这样可以将IO部分与策略解耦。

tianxin1860 replied:
恩，策略和 IO 解耦比较好，我改一下。

tianxin1860 replied:
已解耦，thanks.

Commit:
Merge branch 'develop' into few_shot

Commit:
1. decouple evaluate and train

Commit:
1.add FewCLUE datasets(9/9)
2.implement p-tuning strategy by transform_function
3.unify train_script beteween `chid` task and other 8 tasks of FewCLUE

Commit:
add README.md

ZeyuChen reviewed (approved):
LGTM
